File: dotnet/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example1/CargoAircraft.cs
namespace MediatorDesignPattern.Example1;

/// <summary>
/// Concrete Colleague - Cargo aircraft
/// </summary>
public class CargoAircraft : Aircraft
{
    public CargoAircraft(IMediator mediator, string callSign) 
        : base(mediator, callSign, "Cargo")
    {
    }

    public override void Send(string message)
    {
        Console.WriteLine($"[{CallSign}] Sending: {message}");
        Mediator.SendMessage(message, this);
    }

    public override void Receive(string message)
    {
        Console.WriteLine($"[{CallSign}] Received: {message}");
    }

    public override void RequestLanding()
    {
        Console.WriteLine($"[{CallSign}] Requesting landing clearance...");
        Mediator.RequestLanding(this);
    }

    public override void RequestTakeoff()
    {
        Console.WriteLine($"[{CallSign}] Requesting takeoff clearance...");
        Mediator.RequestTakeoff(this);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example1/src/CargoAircraft.ts
import { Aircraft } from './Aircraft';
import { IMediator } from './IMediator';

export class CargoAircraft extends Aircraft {
  constructor(mediator: IMediator, callSign: string) {
    super(mediator, callSign, 'Cargo');
  }

  send(message: string): void {
    console.log(`[${this.callSign}] Sending: ${message}`);
    this.mediator.sendMessage(message, this);
  }

  receive(message: string): void {
    console.log(`[${this.callSign}] Received: ${message}`);
  }

  requestLanding(): void {
    console.log(`[${this.callSign}] Requesting landing clearance...`);
    this.mediator.requestLanding(this);
  }

  requestTakeoff(): void {
    console.log(`[${this.callSign}] Requesting takeoff clearance...`);
    this.mediator.requestTakeoff(this);
  }
}
